watchdog.go

```diff
@@ -26,239 +76,316 @@ var (
+			}
 
-func watch() {
-	var (
-		lk          sync.Mutex // guards gcTriggered channel, which is drained and flipped to nil on closure.
-		gcTriggered = make(chan struct{}, 16)
+			// calculate how much to set GOGC to honour the next trigger point.
+			currGOGC = int(((float64(next) / float64(heapMarked)) - float64(1)) * 100)
+			if currGOGC >= originalGOGC {
+				Logger.Infof("heap watchdog: requested GOGC percent higher than default; capping at default; requested: %d; default: %d", currGOGC, originalGOGC)
+				currGOGC = originalGOGC
+			} else {
+				if currGOGC < 1 {
+					currGOGC = 1
+				}
+				Logger.Infof("heap watchdog: setting GOGC percent: %d", currGOGC)
+			}
 
-		memstats runtime.MemStats
-		sysmem   gosigar.Mem
-		config   = watchdog.config
-	)
+			debug.SetGCPercent(currGOGC)
 
-	// this non-zero sized struct is used as a sentinel to detect when a GC
-	// run has finished, by setting and resetting a finalizer on it.
-	type sentinel struct{ a *int }
-	var sentinelObj sentinel
-	var finalizer func(o *sentinel)
-	finalizer = func(o *sentinel) {
-		lk.Lock()
-		defer lk.Unlock()
-		select {
-		case gcTriggered <- struct{}{}:
-		default:
-			config.Logger.Warnf("failed to queue gc trigger; channel backlogged")
+			memstatsFn(&memstats)
+			Logger.Infof("heap watchdog stats: heap_alloc: %d, heap_marked: %d, next_gc: %d, policy_next_gc: %d, gogc: %d",
+				memstats.HeapAlloc, heapMarked, memstats.NextGC, next, currGOGC)
 		}
-		runtime.SetFinalizer(o, finalizer)
-	}
-	finalizer(&sentinelObj)
+	}()
 
-	defer watchdog.wg.Done()
-	for {
-		var eventIsGc bool
-		select {
-		case <-Clock.After(config.Resolution):
-			// exit select.
+	return nil, stop
+}
 
-		case <-gcTriggered:
-			eventIsGc = true
-			// exit select.
+// SystemDriven starts a singleton system-driven watchdog.
+//
+// The system-driven watchdog keeps a threshold, above which GC will be forced.
+// The watchdog polls the system utilization at the specified frequency. When
+// the actual utilization exceeds the threshold, a GC is forced.
+//
+// This threshold is calculated by querying the policy every time that GC runs,
+// either triggered by the runtime, or forced by us.
+func SystemDriven(limit uint64, frequency time.Duration, policyCtor PolicyCtor) (err error, stopFn func()) {
+	if !atomic.CompareAndSwapInt32(&_watchdog.state, stateUnstarted, stateRunning) {
+		return ErrAlreadyStarted, nil
+	}
 
-		case <-watchdog.closing:
-			runtime.SetFinalizer(&sentinelObj, nil) // clear the sentinel finalizer.
+	if limit == 0 {
+		limit, err = determineLimit(false)
+		if err != nil {
+			return err, nil
+		}
+	}
 
-			lk.Lock()
-			ch := gcTriggered
-			gcTriggered = nil
-			lk.Unlock()
+	policy, err := policyCtor(limit)
+	if err != nil {
+		return fmt.Errorf("failed to construct policy with limit %d: %w", limit, err), nil
+	}
 
-			// close and drain
-			close(ch)
-			for range ch {
-			}
-			return
+	_watchdog.scope = UtilizationSystem
+	_watchdog.closing = make(chan struct{})
```

Comment:
maybe we can deduplicate this code?

Reply:
Maybe, but it's only a few lines, will make the code less readable, and the LoC will not decrease.